crates/runix/src/store_path.rs

```diff
@@ -0,0 +1,264 @@
+    ///
+    /// let inside_out_path = "/nix/store/7rjqb838snvvxcmpvck1smfxhkwzqal5-python3-3.10.10/bin/python";
+    /// let path = StorePath::from_path(&inside_out_path).unwrap();
+    /// assert_eq!(path.package_path(), Some(Path::new("bin/python")));
+    /// ```
+    pub fn package_path(&self) -> Option<&Path> {
+        self.package_path.as_deref()
+    }
+
+    /// mutable reference to path of a file or directory inside the package
+    ///
+    /// see also: [`StorePath::package_path`]
+    ///
+    /// ```ignore
+    /// /nix/store/7rjqb838snvvxcmpvck1smfxhkwzqal5-python3-3.10.10/bin/python
+    ///                                                            ^^^^^^^^^^^
+    /// ```
+    pub fn package_path_mut(&mut self) -> &mut Option<PathBuf> {
+        &mut self.package_path
+    }
+
+    /// Combine components of store path into a native path type
+    ///
+    /// If parsed from a path, should return an equivalent path
+    ///
+    /// ```
+    /// # use std::path::Path;
+    /// # use runix::store_path::{StorePath, STORE_PREFIX};
+    ///
+    /// let path = Path::new("/nix/store/7rjqb838snvvxcmpvck1smfxhkwzqal5-python3-3.10.10/bin/python");
+    /// assert_eq!(StorePath::from_path(&path).unwrap().as_path(), path);
+    /// ```
+    pub fn as_path(&self) -> PathBuf {
+        if let Some(ref path) = self.package_path {
+            self.out_path().join(path)
+        } else {
+            self.out_path()
+        }
+    }
+
+    /// Try parsing a store path from a pathbuf
+    ///
+    /// Will return an error
+    ///
+    /// * if the path is relative and does not resolve to a store path
+    /// * if the path's prefix does not equal the [STORE_PREFIX]
+    /// * if the path contains '..' components
+    /// * if the path only contains the [STORE_PREFIX]
+    /// ```
+    /// # use runix::store_path::{StorePath, StorePathError, STORE_PREFIX};
+    ///
+    /// assert!(matches!(
+    ///     StorePath::from_path("./relative"),
+    ///     Err(StorePathError::RelativePath(_, _))
+    /// ));
+    ///
+    /// assert!(matches!(
+    ///     StorePath::from_path("/var/nix/store/7rjqb838snvvxcmpvck1smfxhkwzqal5-python3-3.10.10"),
+    ///     Err(StorePathError::NotAStorePath(_))
+    /// ));
+    ///
+    /// assert!(matches!(
+    ///     StorePath::from_path(
+    ///         "/nix/store/7rjqb838snvvxcmpvck1smfxhkwzqal5-python3-3.10.10/bin/../lib"
+    ///     ),
+    ///     Err(StorePathError::RelativeComponent(_))
+    /// ));
+    ///
+    /// assert!(matches!(
+    ///     StorePath::from_path("/nix/store/"),
+    ///     Err(StorePathError::NoPackage(_))
+    /// ));
+    /// ```
+    pub fn from_path(path: impl AsRef<Path>) -> Result<Self, StorePathError> {
+        Self::try_from(path.as_ref().to_path_buf())
+    }
+
+    /// Create a store path given the out_path dir only,
+    /// substituting [STORE_PREFIX].
+    ///
+    /// Likely not the method you are looking for, try [`StorePath::from_path`]
```

Comment:
Should it just be private?

Reply:
removed it, I dont think this is everreally useful, if you need low level access there is the `new_uncehcked` method